# 01 · Carga y exploración inicial

Carga de los dos datasets (`bank-additional.csv` y `customer-details.xlsx`) y diagnóstico de su calidad: columnas, tipos, nulos, duplicados y formato de los valores. **En este notebook no se modifica ningún dato.**

In [1]:
# Rutas y carga de los datos

import pandas as pd

ruta_bank = "../data/raw/bank-additional.csv"
ruta_customer = "../data/raw/customer-details.xlsx"


In [2]:
df_bank = pd.read_csv(ruta_bank)
hojas_customer = pd.read_excel(ruta_customer, sheet_name=None)

print(type(df_bank))
print(type(hojas_customer))

<class 'pandas.DataFrame'>
<class 'dict'>


In [3]:
# Tamaño de los datos

for nombre, hoja in hojas_customer.items():
    print(nombre, hoja.shape)
print(df_bank.shape)

total_customer = sum(hoja.shape[0] for hoja in hojas_customer.values())
print(total_customer)



2012 (20115, 7)
2013 (8965, 7)
2014 (14090, 7)
(43000, 24)
43170


In [4]:
# Columnas: comparar con el enunciado

print(hojas_customer["2012"].columns)
print(df_bank.columns)

Index(['Unnamed: 0', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer',
       'NumWebVisitsMonth', 'ID'],
      dtype='str')
Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'id_'],
      dtype='str')


In [5]:
# Revisar las coordenadas

print(df_bank[["latitude", "longitude"]].describe())

           latitude     longitude
count  43000.000000  43000.000000
mean      36.856697    -95.939067
std        7.225948     16.752282
min       24.396000   -124.997000
25%       30.614750   -110.494250
50%       36.761000    -95.899500
75%       43.113250    -81.427750
max       49.384000    -66.937000


In [6]:
# Formato de la fecha

print(df_bank["date"].head(10))

0         2-agosto-2019
1    14-septiembre-2016
2       15-febrero-2019
3     29-noviembre-2015
4         29-enero-2017
5    26-septiembre-2015
6         25-enero-2019
7    25-septiembre-2017
8         18-enero-2018
9      2-noviembre-2016
Name: date, dtype: str


In [7]:
# Tipos de dato

df_bank.info()

<class 'pandas.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  str    
 3   marital         42915 non-null  str    
 4   education       41193 non-null  str    
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  str    
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  str    
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  str    
 16  cons.conf.idx   43000 non-null  str    
 17  euribor3m       33744 non-null  str    
 1

In [8]:
# Nulos

nulos = df_bank.isnull().sum()
print(nulos[nulos > 0])

porc_nulos = (nulos[nulos > 0] / len(df_bank) * 100).round(2)
print(porc_nulos.sort_values(ascending=False))

age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
cons.price.idx     471
euribor3m         9256
date               248
dtype: int64
euribor3m         21.53
default           20.89
age               11.91
education          4.20
loan               2.39
housing            2.39
cons.price.idx     1.10
job                0.80
date               0.58
marital            0.20
dtype: float64


In [9]:
# Números leídos como texto

cols_economicas = ["emp.var.rate", "cons.price.idx", "cons.conf.idx", "euribor3m", "nr.employed"]
print(df_bank[cols_economicas].head())
print(df_bank["nr.employed"].unique())

   emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed
0           1.1         93,994         -36,4     4,857        5191
1           1.1         93,994         -36,4       NaN        5191
2           1.1         93,994         -36,4     4,857        5191
3           1.1         93,994         -36,4       NaN        5191
4           1.1         93,994         -36,4       NaN        5191
<StringArray>
[  '5191', '5228,1', '5195,8', '5176,3', '5099,1', '5076,2', '5017,5',
 '5023,5', '5008,7', '4991,6', '4963,6']
Length: 11, dtype: str


In [10]:
# Cómo están escritas las categorías

cols_texto = ["job", "marital", "education", "contact", "poutcome", "y"]
for col in cols_texto:
    print(df_bank[col].value_counts(dropna=False))
    print("-" * 40)

job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
NaN                345
Name: count, dtype: int64
----------------------------------------
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
NaN            85
Name: count, dtype: int64
----------------------------------------
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
NaN                     1807
illiterate                18
Name: count, dtype: int64
----------------------------------------
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
----------------------------------------
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count,

In [11]:
# Duplicados y revisión del Excel

print("Filas duplicadas:", df_bank.duplicated().sum())
print("IDs duplicados:", df_bank["id_"].duplicated().sum())

for nombre, hoja in hojas_customer.items():
    print(nombre)
    print("Nulos:", hoja.isnull().sum().sum())
    print(hoja.dtypes)
    print("-" * 40)


Filas duplicadas: 0
IDs duplicados: 0
2012
Nulos: 0
Unnamed: 0                    int64
Income                        int64
Kidhome                       int64
Teenhome                      int64
Dt_Customer          datetime64[us]
NumWebVisitsMonth             int64
ID                              str
dtype: object
----------------------------------------
2013
Nulos: 0
Unnamed: 0                    int64
Income                        int64
Kidhome                       int64
Teenhome                      int64
Dt_Customer          datetime64[us]
NumWebVisitsMonth             int64
ID                              str
dtype: object
----------------------------------------
2014
Nulos: 0
Unnamed: 0                    int64
Income                        int64
Kidhome                       int64
Teenhome                      int64
Dt_Customer          datetime64[us]
NumWebVisitsMonth             int64
ID                              str
dtype: object
----------------------------------------

## Diario de hallazgos

| Dataset | Columna | Problema | Decisión |
|---|---|---|---|
| customer-details | Unnamed: 0 | Índice guardado como columna | Eliminar |
| customer-details | (filas) | 170 clientes más que llamadas | Comprobar al unir |
| customer-details | 3 hojas | Mismas columnas repartidas en 3 tablas | Unirlas con pd.concat sin perder el año |
| bank | Unnamed: 0 | Índice guardado como columna | Eliminar |
| bank | latitude, longitude | Coordenadas de EE. UU. (lat 24–49, lon -125 a -67), no de Portugal | Eliminar |
| bank | contact_month, contact_year | No existen | Crearlas a partir de date |
| bank | date | Texto, no fecha; mes escrito en español (2-agosto-2019) | Traducir meses a número y convertir a fecha |
| bank | cons.price.idx, cons.conf.idx, euribor3m, nr.employed | Leídas como texto: decimal con coma (93,994) | Cambiar coma por punto y convertir a número |
| bank | age, job, marital, education, default, housing, loan, cons.price.idx, euribor3m, date | Tienen nulos (de 0,2 % a 21,5 %) | Decidir según el % y el tipo de cada una |
| bank | default, housing, loan | Son 0/1 pero salen como float por los nulos | Pasar a entero tras tratar los nulos |
| bank | marital, poutcome | En mayúsculas; el resto en minúsculas | Pasar todo a minúsculas |
| bank | job, education | Puntos en los valores (admin., university.degree) | Limpiar puntos |
| bank | y | Texto yes/no | Convertir a 1/0 para calcular tasas |


## Primeras Observaciones

- Solo el **11,3 %** de las llamadas acaba en contratación (4.844 de 43.000): una de cada nueve.
- El **86 %** de los clientes no había sido contactado en campañas anteriores (poutcome = nonexistent).
- No hay filas ni IDs duplicados en el CSV, y el Excel no tiene nulos.
- Las llamadas están repartidas casi por igual entre 2015 y 2019 (unas 8.500 por año).
